cpp/src/utilities/stream_pool.cpp

```diff
@@ -161,36 +161,45 @@ cuda_stream_pool* create_global_cuda_stream_pool()
   return new rmm_cuda_stream_pool();
 }
 
+// FIXME: these will be available in rmm soon
+inline int get_num_cuda_devices()
+{
+  rmm::cuda_device_id::value_type num_dev{};
+  CUDF_CUDA_TRY(cudaGetDeviceCount(&num_dev));
+  return num_dev;
+}
+
+rmm::cuda_device_id get_current_cuda_device()
+{
+  int device_id;
+  CUDF_CUDA_TRY(cudaGetDevice(&device_id));
+  return rmm::cuda_device_id{device_id};
+}
+
 /**
  * @brief RAII struct to wrap a cuda event and ensure its proper destruction.
  */
 struct cuda_event {
-  cuda_event()
-    : e_{[]() {
-        cudaEvent_t event;
-        CUDF_CUDA_TRY(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
-        return event;
-      }()}
-  {
-  }
+  cuda_event() { CUDF_CUDA_TRY(cudaEventCreateWithFlags(&e_, cudaEventDisableTiming)); }
+  virtual ~cuda_event() { CUDF_ASSERT_CUDA_SUCCESS(cudaEventDestroy(e_)); }
 
-  operator cudaEvent_t() { return e_.get(); }
+  operator cudaEvent_t() { return e_; }
 
  private:
-  struct deleter {
-    using pointer = cudaEvent_t;
-    auto operator()(cudaEvent_t e) { CUDF_ASSERT_CUDA_SUCCESS(cudaEventDestroy(e)); }
-  };
-  std::unique_ptr<cudaEvent_t, deleter> e_;
+  cudaEvent_t e_;
 };
 
 /**
  * @brief Returns a cudaEvent_t for the current thread.
  */
 cudaEvent_t event_for_thread()
 {
-  thread_local cuda_event thread_event;
-  return thread_event;
+  thread_local std::vector<std::unique_ptr<cuda_event>> thread_events(get_num_cuda_devices());
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏ Curious to discuss `unique` vs. `shared` here. I used `shared` in my concurrent RMM PR, but I don't feel like an expert about C++ smart pointers.

I *think* `unique_ptr` is right here because the ptr itself is never copied. In my code I used `shared_ptr` because I was thinking "multiple threads, multiple copies", but really there's a separate ptr (and event) per thread in `thread_local`, not copies of the same pointer. So I could change my code to `unique_ptr`. Am I right?

Reply:
I think that sounds right.

Reply:
IIUC, you return a {cudaStream_t, cudaEvent_t} tuple, not the smart pointer itself. So yes, I _think_ you could just use `unique_ptr` as well.  @vuule or @bdice , do you care to comment?

Reply:
Yeah, no need for `shared_ptr`, since the pointer itself is never copied. The thread local vector is the sole owner of the events.

Reply:
Actually, when I went to change to `unique_ptr` in my code here: https://github.com/rapidsai/rmm/pull/1333/files#diff-0d9e7691f84fe22b13d13977fe225d29f7824feee8e5b5763d98d8569b87634eR295-R304

I realized that I need `shared_ptr` because I actually stash a copy of the shared pointer in another container. The purpose of that copy is because the events may still be needed when a block is deallocated, *after* the thread exits and the TLS is freed.

Thinking about whether you need to do the same in this PR, I don't think so. However it made me wonder why you are using TLS at all. Is it just in case the stream passed to `fork` is the PTDS?  Could you just have a static event for each device? That would mean if you create multiple stream pools for a single device they would share an event, but hopefully nobody will do that...

Reply:
I'm sorry, I'm not thoroughly steeped in cuda terminology yet. PTDS I get, but what is TLS?

Reply:
Yes, Thread Local Storage. That's C++, not CUDA.  Sorry for the unexplained TLA (Three Letter Acronym).

Reply:
Thanks, at least I guessed correctly 😅 I had to go back to look at some offline conversations with @vuule, but the reason for the TLS was a fear that if two threads called either fork or join at the same time, there could be a condition where thread 1 records an event on stream 1, then thread 2 records an event on stream 2, but both calls use the same static event. My understanding of the docs was that the second call to cudaEventRecord would overwrite the state on the event, so calls to cudaStreamWaitEvent from thread 1 would actually be waiting on stream 2. @vuule wrote a quick per-thread event store, but then @jrhemstad pointed out we could just use TLS to get an event per thread.

Reply:
OK yes this is similar to the reason for the TLS in RMM's stream_ordered_memory_resource.  And I don't think you need the extra shared copy that we need there because the event is only used within the scope of one function. 👍